src/persist-client/src/iter.rs

```diff
@@ -0,0 +1,701 @@
+    part_key: PartialBatchKey,
+    part_desc: Description<T>,
+}
+
+impl<T: Codec64 + Timestamp + Lattice> FetchData<T> {
+    async fn fetch(self) -> anyhow::Result<EncodedPart<T>> {
+        let Self {
+            shard_id,
+            blob,
+            metrics,
+            read_metrics,
+            shard_metrics,
+            part_key,
+            part_desc,
+        } = self;
+        metrics.compaction.parts_prefetched.inc();
+        fetch_batch_part(
+            &shard_id,
+            &*blob,
+            &metrics,
+            &shard_metrics,
+            read_metrics(&metrics.read),
+            &part_key,
+            &part_desc,
+        )
+        .await
+    }
+}
+
+#[derive(Debug)]
+pub(crate) enum ConsolidationPart<T, D> {
+    Queued {
+        data: Option<FetchData<T>>,
+    },
+    Prefetched {
+        handle: JoinHandle<anyhow::Result<EncodedPart<T>>>,
+        metrics: Arc<Metrics>,
+    },
+    Encoded {
+        part: EncodedPart<T>,
+        cursor: Cursor,
+    },
+    Sorted {
+        data: Vec<((Vec<u8>, Vec<u8>), T, D)>,
+        index: usize,
+    },
+}
+
+impl<'a, T: Timestamp + Codec64 + Lattice, D: Codec64 + Semigroup> ConsolidationPart<T, D> {
+    pub(crate) fn from_encoded(part: EncodedPart<T>, as_of: &Antichain<T>) -> Self {
+        let mut cursor = Cursor::default();
+        if part.maybe_user_part() {
+            Self::from_iter(ConsolidationPartIter::encoded(&part, &mut cursor, as_of))
+        } else {
+            ConsolidationPart::Encoded { part, cursor }
+        }
+    }
+
+    pub(crate) fn from_iter(data: impl IntoIterator<Item = TupleRef<'a, T, D>>) -> Self
+    where
+        D: Semigroup,
+    {
+        let mut data: Vec<_> = data.into_iter().map(clone_tuple).collect();
+        consolidate_updates(&mut data);
+        Self::Sorted { data, index: 0 }
+    }
+
+    async fn join(&mut self, as_of: &Antichain<T>) -> anyhow::Result<()> {
+        match self {
+            ConsolidationPart::Queued { data } => {
+                let data = data.take().expect("unfetched");
+                data.metrics.compaction.parts_waited.inc();
+                *self = Self::from_encoded(data.fetch().await?, as_of);
+            }
+            ConsolidationPart::Prefetched { handle, metrics } => {
+                metrics.compaction.parts_waited.inc();
+                *self = Self::from_encoded(handle.await??, as_of);
+            }
+            ConsolidationPart::Encoded { .. } | ConsolidationPart::Sorted { .. } => {}
+        }
+        Ok(())
+    }
+
+    pub(crate) fn is_empty(&self) -> bool {
+        match self {
+            ConsolidationPart::Encoded { part, cursor, .. } => cursor.clone().peek(part).is_none(),
+            ConsolidationPart::Sorted { data, index } => data.len() <= *index,
+            ConsolidationPart::Queued { .. } | ConsolidationPart::Prefetched { .. } => false,
+        }
+    }
+}
+
+#[derive(Debug)]
+pub(crate) struct Consolidator<T, D> {
```

Comment:
> is that ordering required or can we e.g. enqueue some new runs after we've started iterating?

IIRC that would be fine as long as the new run always had keys larger than any that had been returned by `next`. On the other hand I don't have a nice way to check that automatically yet, and I don't really see where it would be useful.

I'm going to look into adding some static enforcement like you suggest, and if that ends up being too painful I will do a prominent rustdoc!